### ISSUE TITLE
Gap in titlebar when some windows use higher glyphs

### ISSUE THREAD

**Issue by minus7** [swaywm/sway#2455]
![screenshot](https://user-images.githubusercontent.com/419104/44002577-23efa286-9e45-11e8-86ea-da2010f3e9e9.png)

Reproduce with: `urxvt -title 'Hello 世界' -e /usr/bin/sleep 20h`

Maybe related to #1936?

**Comment by minus7** [comment]
`pango_layout_get_pixel_size` is used in the background, which should return the logical extents of the text, which should, as far as I understand it, include spacing for characters that extend beyond the baseline, etc. My guess is that CJK characters use a different font and cause the height to change. To make the  titlebar height stable, it'd have to always use the maximum height that could ever appear

**Comment by RyanDwyer** [comment]
This was an issue a while ago and I made a fix for it but it appears to have regressed. The fix was to fill in the gap below the title. If you figure out why that's no longer working then you'll probably find that it's easy to fix.

**Comment by minus7** [comment]
Filling the gap alone has two problems: The text will not be centered and titlebars will change height when a window with tall characters appears/disappears

**Comment by RyanDwyer** [comment]
The changing height for all containers is intentional. I can't say I'm a fan of it, but it's needed so that the title bars are the same height for tabbed containers.

If we could make it so it only changes the height for the glyph container (and if it's in a tabbed container, the siblings too) and leave the others at their normal height then that'd be cool. We could remove the whole recalculating font height thing in the config then.

**Comment by SirCmpwn** [comment]
No, we should have all window titles be at the same height, as a function of the tallest title text currently in use. We just need to center it and fill in the extra space for titles which are shorter than this.

**Comment by minus7** [comment]
I'd prefer having it always use the tallest possible height so that my media player or web browser suddenly changing title does not change all titlebars

**Comment by SirCmpwn** [comment]
"The tallest possible height" is not a thing we can know without trying to rasterize all valid unicode characters.

**Comment by SirCmpwn** [comment]
We can do the tallest height *we've seen yet*, perhaps?

**Comment by SirCmpwn** [comment]
>"The tallest possible height" is not a thing we can know without trying to rasterize all valid unicode characters.

Actually you can use combining characters to make text arbitrarily tall

**Comment by RyanDwyer** [comment]
> I'd prefer having it always use the tallest possible height so that my media player or web browser suddenly changing title does not change all titlebars

-1 to this. If I open a YouTube video or something with a unicode character, I don't want all my titlebars to be unnecessarily large until the next time I restart sway.

**Comment by minus7** [comment]
Just tried a title with combined characters; it didn't become any larger than in the screenshot above

**Comment by SirCmpwn** [comment]
>Just tried a title with combined characters; it didn't become any larger than in the screenshot above

sounds like a bug tbh

**Comment by SirCmpwn** [comment]
Is this still a problem? I cannot reproduce.

**Comment by minus7** [comment]
Yep, still pretty much the same.
